filegenerator/file_generator.go

```diff
@@ -0,0 +1,393 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: William Haack (will@cockroachlabs.com)
+
+// The purpose of filegenerator is to provide the tools for the automatic
+// deployment of JSON files that are used on the benchviz web server
+// as the data to be displayed. RenderHistoricalBenchmarkResults
+// creates a BenchPackages struct that has all the data needed to
+// generate most of the JSON files. The various other methods use
+// the BenchPackages created by RenderHistoricalBenchmarkResults
+// in order to create the JSON files.
+
+package filegenerator
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	deployRootEnv    string = "BENCHDEPLOY"
+	benchSamplesEnv  string = "BENCHSAMPLES"
+	awsBucketNameEnv string = "AWSBUCKETNAME"
+	cockroachRootEnv string = "MYROOT"
+)
+
+// BenchPackages is a mapping of package names to test names, which in turn map
+// to their historical data.
+type BenchPackages map[string]BenchTestMap
+
+// BenchTestMap is a mapping of test names to dates, which in turn map to the
+// results of the benchmark tests for that given date.
+type BenchTestMap map[string]BenchResults
+
+// BenchResults map dates to the benchmark test data for that date.
+type BenchResults map[string]BenchStats
+
+// BenchStats is a struct containing the various values obtained from running
+// a benchmark test with the -benchmem parameter. The variable names are kept
+// short to keep the JSON files small.
+type BenchStats struct {
+	N int     //nsop
+	A int     //alop
+	B int     //bop
+	M float64 //mbps
+}
+
+// GeometricMeanData is a struct that contains the means of the benchmark
+// stats for all the tests in a given package.
+type GeometricMeanData struct {
+	NMean float64
+	AMean float64
+	BMean float64
+	MMean float64
+	Date  string
+}
+
+func isValidTestLine(line string) bool {
+	matched, err := regexp.MatchString("FAIL", line)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return len(line) >= 9 && line[0:9] == "Benchmark" && !matched
```

Comment:
you should add a comment somewhere describing or linking to the expected input format. without that, this code makes only superficial sense to me